### ISSUE TITLE
'ERR EXEC without MULTI' when using transactions with large number of requests

### ISSUE THREAD

**Issue by andre2i** [vert-x3/vertx-redis-client#87]
### Version

* vert.x core: 3.4.1
* vert.x redis client: 3.4.1

### Context

I was implementing a new data model for the service's underlying cache and used transactions for adding new keys.

After looking at the examples [here](https://github.com/vert-x3/vertx-redis-client/blob/master/src/main/java/examples/Examples.java#L72) and [here](https://github.com/vert-x3/vertx-redis-client/blob/573fd98b28892f01e5d9d5bf2a14a5281bf291b8/src/test/java/io/vertx/test/redis/RedisClientTestBase.java#L685).

I implemented something like the following:

```
        RedisTransaction transaction = redisClient.transaction();
        transaction.multi(multi -> {
            addSomeKeys(transaction);
            addSomeOtherKeysInOtherPlaces(transaction);
            transaction.exec(exec -> {
                if (exec.succeeded()) {
                    handler.handle(Future.succeededFuture());
                } else {
                    handler.handle(Future.failedFuture(exec.cause()));
                }
            });
        });
```
which works well in the integration tests and low-key manual testing. However when subjecting it to a load test the service logs are flooded with `ERR EXEC without MULTI`. The load is not that big (it starts reproducing after surpassing about 30 rps on my machine). Our service logic requeues the data structure to be added and it is added eventually.

When changing the code to the following pattern:

```
       RedisTransaction transaction = redisClient.transaction();
        transaction.multi(multi -> {
            addSomeKeys(transaction);
            addSomeOtherKeysInOtherPlaces(transaction);
        }).exec(exec -> {
                if (exec.succeeded()) {
                    handler.handle(Future.succeededFuture());
                } else {
                    handler.handle(Future.failedFuture(exec.cause()));
                }
       });
```

Everything works just fine in all scenarios with 0 errors a high RPS. But when looking at the MONITOR output from Redis, It doesn't look like a transaction:

```
**1503601987.997592 [1 127.0.0.1:50172] "MULTI"**
**1503601987.997599 [1 127.0.0.1:50172] "EXEC"**
1503601987.999116 [1 127.0.0.1:50172] "SADD" 
1503601987.999891 [1 127.0.0.1:50172] "SADD" 
1503601988.000368 [1 127.0.0.1:50172] "SADD"
1503601988.000699 [1 127.0.0.1:50172] "SADD" 
1503601988.001755 [1 127.0.0.1:50172] "ZADD" 
1503601988.208249 [1 127.0.0.1:50174] "SINTER"
1503601988.209844 [1 127.0.0.1:50174] "ZRANGE" 
1503601988.210912 [1 127.0.0.1:50169] "FLUSHALL"
**1503601988.214041 [1 127.0.0.1:50172] "MULTI"**
**1503601988.214048 [1 127.0.0.1:50172] "EXEC"**
1503601988.214474 [1 127.0.0.1:50172] "SADD" 
1503601988.214494 [1 127.0.0.1:50172] "ZREMRANGEBYRANK" 
1503601988.214875 [1 127.0.0.1:50172] "SADD"
1503601988.215238 [1 127.0.0.1:50172] "SADD"
1503601988.215622 [1 127.0.0.1:50172] "SADD"
1503601988.215993 [1 127.0.0.1:50172] "ZADD"
1503601988.418723 [1 127.0.0.1:50174] "SINTER"
1503601988.419022 [1 127.0.0.1:50174] "ZRANGE"
1503601988.420029 [1 127.0.0.1:50169] "FLUSHALL"
```
Below is a snippet of the MONITOR output with the transaction executed within .multi:

```
1503604626.549065 [0 127.0.0.1:49440] "MULTI"
1503604626.549490 [0 127.0.0.1:49440] "MULTI"
1503604626.556246 [0 127.0.0.1:49440] "MULTI"
1503604626.556927 [0 127.0.0.1:49440] "MULTI"
1503604626.561319 [0 127.0.0.1:49440] "SADD" 
1503604626.561359 [0 127.0.0.1:49440] "SADD" 
1503604626.561397 [0 127.0.0.1:49440] "SADD" 
1503604626.561434 [0 127.0.0.1:49440] "SADD"
1503604626.561466 [0 127.0.0.1:49440] "SADD" 
1503604626.561506 [0 127.0.0.1:49440] "SADD" 
1503604626.561542 [0 127.0.0.1:49440] "SADD"
1503604626.561582 [0 127.0.0.1:49440] "SADD" 
1503604626.561609 [0 127.0.0.1:49440] "SADD"
1503604626.561644 [0 127.0.0.1:49440] "SADD" 
1503604626.561684 [0 127.0.0.1:49440] "SADD"
1503604626.561721 [0 127.0.0.1:49440] "SADD" 
1503604626.561748 [0 127.0.0.1:49440] "SADD"
1503604626.561783 [0 127.0.0.1:49440] "SADD" 
1503604626.561819 [0 127.0.0.1:49440] "SADD"
1503604626.561851 [0 127.0.0.1:49440] "ZADD" 
1503604626.561982 [0 127.0.0.1:49440] "ZADD"
1503604626.563832 [0 127.0.0.1:49440] "SADD"
1503604626.563883 [0 127.0.0.1:49440] "ZADD"
1503604626.564327 [0 127.0.0.1:49440] "EXEC"
```

### Do you have a reproducer?

Not right now as the load testing setup in which I found this is quite convoluted.

### Extra

Considering that this is happening when the service actually gets concurrent traffic, I suspect this is something either related to my API usage or to the overlapping MULTI commands on the server side, so my questions are:

* before I try to work on a consistent repro, has anyone ever saw this error from this Redis client? I had trouble googling for this particular error in the vert.x context.
* I am assuming that the ordering of commands in the Redis monitor is the order in which they were processed by the server. The assumption is correct, right?
* Do we need to handle the transactional manner of the API within the service logic? I haven't ran into many examples of transactions using the vert.x Redis client but do we need to be mindful and protect against overlapping MULTI requests?
* Should I just abandon this approach and go with EVAL to make sure the server locks between MULTI and EXEC?

**Comment by pmlopes** [comment]
should be fixed by #101 (which is a new proposed API) there is now a `batch` method that allows you to send a batch of commands at once (without interleaves of other sources). Please review that PR.

**Comment by pmlopes** [comment]
Batch is available on 4.0.0 milestone 4
